pkg/repo/repo.go

```diff
@@ -127,36 +159,270 @@ func CreateRepoWithMetadata(ctx context.Context, targets []TargetWithMetadata) (
+	}
+	result = append(result, root...)
+	return result
+}
+
+func pubkeyToTLogInstance(key []byte) (*v1.TransparencyLogInstance, error) {
+	logId := sha256.Sum256(key)
+	der, _ := pem.Decode(key)
+	keyDetails, err := getKeyDetails(der.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1.TransparencyLogInstance{
+		BaseUrl:       "",
+		HashAlgorithm: v1Common.HashAlgorithm_SHA2_256, // TODO: make it possible to change this value
+		PublicKey: &v1Common.PublicKey{
+			RawBytes:   der.Bytes,
+			KeyDetails: keyDetails,
+			ValidFor: &v1Common.TimeRange{
+				Start: timestamppb.New(time.Now()),
+			},
+		},
+		LogId: &v1Common.LogId{
+			KeyId: logId[:],
+		},
+	}, nil
+}
+
+func getKeyDetails(key []byte) (v1Common.PublicKeyDetails, error) {
+	var k any
+	var err1, err2 error
+
+	k, err1 = x509.ParsePKCS1PublicKey(key)
+	if err1 != nil {
+		k, err2 = x509.ParsePKIXPublicKey(key)
+		if err2 != nil {
+			return 0, fmt.Errorf("Can't parse public key with PKCS1 or PKIX: %w, %w", err1, err2)
+		}
+	}
+
+	// borrowed from https://github.com/kommendorkapten/trtool/blob/main/cmd/trtool/app/common.go
+	switch v := k.(type) {
+	case *ecdsa.PublicKey:
+		if v.Curve == elliptic.P256() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P256_SHA_256, nil
+		}
+		if v.Curve == elliptic.P384() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P384_SHA_384, nil
+		}
+		if v.Curve == elliptic.P521() {
+			return v1Common.PublicKeyDetails_PKIX_ECDSA_P521_SHA_512, nil
+		}
+		return 0, errors.New("unsupported elliptic curve")
+	case *rsa.PublicKey:
+		/*
+			NOTE: It is not possible to recognize padding from just the public key alone;
+			we will just assume that the padding used is pkcs1v15
+			if padding == RSAPSS {
+				switch v.Size() * 8 {
+				case 2048:
+					return v1Common.PublicKeyDetails_PKIX_RSA_PSS_2048_SHA256, nil
+				case 3072:
+					return v1Common.PublicKeyDetails_PKIX_RSA_PSS_3072_SHA256, nil
+				case 4096:
+					return v1Common.PublicKeyDetails_PKIX_RSA_PSS_4096_SHA256, nil
+				default:
+					return 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+				}
+			}
+		*/
+		switch v.Size() * 8 {
+		case 2048:
+			return v1Common.PublicKeyDetails_PKIX_RSA_PKCS1V15_2048_SHA256, nil
+		case 3072:
+			return v1Common.PublicKeyDetails_PKIX_RSA_PKCS1V15_3072_SHA256, nil
+		case 4096:
+			return v1Common.PublicKeyDetails_PKIX_RSA_PKCS1V15_4096_SHA256, nil
+		default:
+			return 0, fmt.Errorf("unsupported public modulus %d", v.Size())
+		}
+	case ed25519.PublicKey:
+		return v1Common.PublicKeyDetails_PKIX_ED25519, nil
+	default:
+		return 0, errors.New("unknown public key type")
+	}
+}
+
+func certChainToAuthority(certChainPem []byte) (*v1.CertificateAuthority, error) {
+	var cert *x509.Certificate
+	var err error
+	rest := certChainPem
+	certChain := v1Common.X509CertificateChain{Certificates: []*v1Common.X509Certificate{}}
+
+	// skip potential whitespace at end of file (8 is kinda random, but seems to work fine)
```

Comment:
I'm not sure this is necessary, I believe whitespace will be handled by `pem.Decode`. Would be good to test for that.